src/role_extractor.js

```diff
@@ -21,10 +21,58 @@ roles.extractor.energyBuild = function(room, energy, heal) {
   return energy;
 };
 
-roles.extractor.action = function(creep) {
-  return creep.handleExractor();
-};
+roles.extractor.terminalStorageExchange = function(creep) {
+  var terminal = creep.room.terminal;
+  if (!terminal || !terminal.isActive()) {
+    // TODO kill creep?
+    return ERR_INVALID_TARGET;
+  }
+  var energyInTerminal = terminal.store.energy / terminal.storeCapacity;
+  var totalInTerminal = _.sum(terminal.store) / terminal.storeCapacity;
+  if ((energyInTerminal < 0.5) && (totalInTerminal !== 1)) {
+    return ERR_NOT_ENOUGH_RESOURCES;
+  }
+
+  if (totalInTerminal > 0.5) {
+    // TODO call carry to move energy from
+  }
+  // transferToStructures then decide go to terminal or storage
+  creep.transferToStructures();
+
+  var action = {
+    withdraw: _.sum(creep.carry) / creep.carryCapacity < 0.8,
+    transfer: _.sum(creep.carry) / creep.carryCapacity > 0.3
+  };
+
+  // TODO replace creep.moveTo by moving on path ?
+
+  if (action.withdraw) {
+    if (creep.withdraw(terminal, RESOURCE_ENERGY) !== OK) {
+      creep.moveTo(terminal);
+    }
+  }
 
-roles.extractor.execute = function(creep) {
-  return creep.handleExractor();
+  if (!action.withdraw || action.transfer) {
+    if (creep.transfer(creep.room.storage, RESOURCE_ENERGY) !== OK) {
+      creep.moveTo(creep.room.storage);
```

Comment:
I guess this will run into problems. Especially because the harvester creep is on the path and block the moveTo, because no `ignoreCreeps`. If `ignoreCreeps` is set, it could get stuck at the filler or builder creep.

But it also can work sometimes ...